### ISSUE TITLE
Proposal: Schemas in stable storage

### ISSUE THREAD

**Issue by arschles** [Azure/open-service-broker-azure#171]
There's a schema change going on now, and I believe we want to get to a stable schema soon so we don't have to break it later.

I wonder if it would be worth using protobufs to store data, so that we can mutate the schema in predictable, backward compatible ways going forward.

This is all for the data store, not for the task queue.

We could also use protobufs for task queue payloads, but I'll leave that outside the scope of this issue

cc/ @krancour @jeremyrickard

**Comment by krancour** [comment]
This is an interesting thought. My biggest technical reservation is that we have to mind encryption. I'm not sure how (if at all), individual fields can be encrypted in a protobuf. Note that I deliberately don't want entire records encrypted at rest-- just the (potentially) sensitive parts because access to meta-data about instances and bindings seems valuable for troubleshooting.

Maybe we just need to handle encryption and decryption before/after marshaling/unmarshaling.

Technical concerns aside, I'm also concerned that mainly for lack of previous experience with protobufs (excepting some very minor dabbling with grpc), this doesn't look like a light lift.

I would say that we should explore this if and when we feel ready to spend several weeks on it.

**Comment by krancour** [comment]
I've had time to research protobufs more extensively and reason through a few things. I've stopped short of prototyping this and will save that exercise for when time permits.

Here are some updated thoughts:

This would protect us from some, but not all, schema changes. Let's look at some recent changes we've actually implemented to the model that affected the storage schema to understand what advantages (if any) protobufs can offer us over JSON. For each of the examples below, I'll decompose the changes into two concerns: storage schema changes and model changes.

In one case, we recently renamed an `Instance` field from `ProvisioningContext` to `Details`. (We felt that "instance details" more clearly communicated the purpose of that field.) We changed the schema _and_ the model. Note that we _could_ have elected to leave the schema untouched and simply mapped the old field name  in the schema (`provisioningContext`) to the new field name in the model (`Details`). If we'd already reached a point where changes to the schema were not permissible we would have chosen that. In either case, the lion's share of the effort for this change resulted from the change not to the schema, but to the _model_. Much code that formerly referenced the `ProvisioningContext` field of `Instance` objects had to be updated to reference the `Details` field instead.

Compare this to how we would have managed the same change with protobufs. We'd have edited a `.proto` file and changed the name of the `provisioningContext` field to `details` _without_ touching the numeric index for the field. This would have prevented any breaking change to the schema. An updated model would have been generated from the `.proto` and then, as was the case with JSON, much code that formerly referenced the `ProvisioningContext` field of `Instance` would have needed to be updated to reference the `Details` field instead.

In this case, JSON and protobufs are tied in terms of what they can do to insulate us from schema changes and the effort required to manage changes the model.

Taking another recent case:

We recently took the `Location`, `ResourceGroupe`, and `Tags` fields from the `StandardProvisioningParameters` field of an `Instance` and moved them onto the `Instance`. (i.e. We "flattened" it.) This was a breaking change to the storage schema, and with JSON, there would have been no way to do this while avoiding this breaking change-- unless we wanted to write complex unmarshalling code. Once again, the lion's share of the effort went into adapting existing code to the model changes.

Compare this to how we would have managed the same change with protobufs... and there's almost no difference. As with JSON, this change could not be accomplished with protobufs _without_ a breaking change to the storage schema. Protobufs rely on immutable numeric indices for all fields. This makes it easy to add, delete, or rename fields, as long as the established indices are never modified.  But those indices are scoped to individual message types. This means it's not possible to take fields `n` and `m` from an embedded `StandardProvisioningParameters` and transplant them onto the `Instance`. Again, as with JSON, the very best we could hope to do to avoid breaking the schema in this scenario is to leave the old fields intact and write complex "unmarshaling" code to map the old schema onto model. There's an inherent disadvantage here, however, that the model generated from the updated `.proto` would have `Location`, `ResourceGroup`, and `Tag` fields on the `Instance` _and_ embedded in the (old) `StandardProvisioningParameters` field. This results in a model that's difficult to work with and would probably lead to us never refactoring the model at all in scenarios such as this. Over time, that's certain to result in increased accrual of tech debt.

In this case, I feel JSON edges out protobufs by a non-trivial margin.

Lastly, I'll circle back to a concern that I'd already expressed-- the degree to which records will be human readable. I'd forgotten that protobufs are (only or by default?) binary. This means any concerns of encryption aside (with JSON, we encrypt certain fields; with protobufs, we'd encrypt the whole thing) anything stored in a protobuf is inherently unreadable by a human-- and that seems as if it will complicate troubleshooting substantially.

I have one tiny glimmer of hope that protobufs might allow us to drop some of our existing code. There's some polymorphic behavior in our storage. Every `Instance` is composed of service-specific `ProvisioningParameters` and `Details`. JSON does not support this well. Currently, what we must do is unmarshal an `Instance` to learn its `serviceID`. Once that is known, we can obtain the service from the catalog, and from that service's `ServiceManager`, we can obtain service-specific implementation of `ProvisioningParameters` and `Details`. Once we have these, we can take a second pass at unmarshaling to ensure the fields in those service-specific types are populated from storage. (Something similar goes on for `Binding`s as well.) There's an "any" type in protobufs that might support this well-- but as I understand it, support for that is a WIP for most languages. I am not clear on how well Go handles this and I believe only a PoC will reveal that to us.

Bottom line-- the advantages of protobuf over JSON for storage are tenuous at best.

I'm going to mark this as low priority and propose that we revisit it shortly before considering graduating from alpha into beta. In the meantime, there are higher value things we can work on.

**Comment by arschles** [comment]
@krancour Awesome write up & research :+1:

Also:

>I'm going to mark this as low priority and propose that we revisit it shortly before considering graduating from alpha into beta. In the meantime, there are higher value things we can work on.

I'm cool with this. Happy to pick up discussion on the protobuf features, etc... sometime later if it's appropriate. I have some experience here.
